{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#import lxml, html5lib\n",
    "import dash\n",
    "from dash.dependencies import Input, Output\n",
    "import dash_core_components as dcc\n",
    "import dash_html_components as html\n",
    "import dash_table\n",
    "\n",
    "from zipfile import ZipFile\n",
    "from glob import glob\n",
    "\n",
    "from time import sleep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "batches = glob(\"Batches/*.zip\")\n",
    "availableBatches = [batch for batch in batches]\n",
    "\n",
    "omim_map = pd.read_csv('Databases/OMIM_map.txt', sep = '\\t')\n",
    "\n",
    "MGI_IDs = pd.read_csv('Databases/MGI_IDs.csv', dtype = str)\n",
    "MGI_IDs = dict(zip(MGI_IDs['Symbol'], MGI_IDs['MGI_ID']))\n",
    "\n",
    "hpo = pd.read_csv('Databases/HPO_dict.txt', sep = '\\t')\n",
    "hpo = hpo[hpo['Associated Genes'].str.len() > 1]\n",
    "\n",
    "refSeqSummaries = pd.read_csv('Databases/RefSeqSummaries.tsv', sep = '\\t', dtype = str)\n",
    "\n",
    "default_style = {'width' : '57%', 'height' : '100%', 'margin' : '10px', 'font-family' : 'gisha', 'marginLeft' : 'auto', 'marginRight' : 'auto', 'textAlign' : 'center'}\n",
    "hide = {'display' : 'None'}\n",
    "full_display = {'width' : '100%', 'height' : '100%', 'margin' : '10px', 'font-family' : 'gisha', 'textAlign' : 'center'}\n",
    "info_style = {'width' : '100%', 'height' : '100%', 'margin' : '10px', 'font-family' : 'gisha', 'marginLeft' : 'auto', 'marginRight' : 'auto', 'textAlign' : 'left'}\n",
    "break_line = html.Hr(style={'height' : '4px', 'width' : '60%', 'color' : '#111111','display' : 'inline-block', 'marginLeft':'auto', 'marginRight':'auto'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateHeatTable(gene, Chromosome, Position):\n",
    "    sleep(2)\n",
    "    children = []\n",
    "    try:\n",
    "        url = 'https://www.ncbi.nlm.nih.gov/clinvar?term=(' + gene + '%5BGene%20Name%5D)%20AND%20pathogenic'\n",
    "        df = pd.read_html(url)[0]\n",
    "        if df.shape[0] > 100:\n",
    "            children.append(html.P('Too many pathogenic variants for this gene!ðŸ˜µ'))\n",
    "        elif df.shape[0] == 0:\n",
    "            children.append(html.P('Could not find pathogenic variants for this gene in ClinVarðŸ¤· ðŸ¤·â€â™‚ï¸'))\n",
    "        else:\n",
    "            del df['VariationLocation']\n",
    "            del df['Gene(s)']\n",
    "            del df['Clinical significance (Last reviewed)']\n",
    "            del df['Review status']\n",
    "            del df['Protein change']\n",
    "            df.fillna('', inplace = True)\n",
    "            def getGRCh38(value):\n",
    "                try:\n",
    "                    location = value.split('GRCh38: Chr')[1]\n",
    "                    if '-' in location:\n",
    "                        location = location.split('-')[0]\n",
    "                    return location\n",
    "                except:\n",
    "                    return 'delete'\n",
    "            def getVariation(value):\n",
    "                try:\n",
    "                    variant = value.replace('GRCh38/hg38 ','').split('GRC')[0]\n",
    "                    return variant\n",
    "                except:\n",
    "                    return 'delete'\n",
    "            df['Coordinates'] = df['VariationLocation.1'].apply(getGRCh38)\n",
    "            df['Variation'] = df['VariationLocation.1'].apply(getVariation)\n",
    "            del df['VariationLocation.1']\n",
    "            df = df[df['Coordinates'] != 'delete']\n",
    "            df = df[df['Variation'] != 'delete']\n",
    "            df['Condition(s)'] = df['Condition(s)'].str.replace('See cases', 'Check link âžœ')\n",
    "            df['Position'] = df['Coordinates'].str.split(':').str[1]\n",
    "            df['Position'] = df['Position'].astype(int)\n",
    "            coordinates = str(Chromosome) + ':' + str(Position)\n",
    "            if coordinates in df['Coordinates']:\n",
    "                False\n",
    "            else:\n",
    "                df = df.append({\n",
    "                    'Coordinates' : coordinates,\n",
    "                    'Position' : int(Position),\n",
    "                    'Condition(s)' : 'ðŸ‘¨â€ðŸ”¬ðŸ‘©â€ðŸ”¬',\n",
    "                    'Accession' : 'Your variant',\n",
    "                    'Variation' : 'ðŸ‘©â€âš•ï¸ ðŸ‘¨â€âš•ï¸'\n",
    "                },\n",
    "            ignore_index=True)\n",
    "            df.sort_values('Position', inplace = True)\n",
    "            if df['Position'].max() - int(Position) >= int(Position) - df['Position'].min():\n",
    "                maxSpace = df['Position'].max() - int(Position)\n",
    "            else:\n",
    "                maxSpace = int(Position) - df['Position'].min()\n",
    "            df['gradient'] = (1 - abs(int(Position) - df['Position']) / maxSpace) * 255 ** 0.7\n",
    "            df['gradient'] = df['gradient'].astype(int)\n",
    "            df.at[df.index[df['Coordinates'] == coordinates].tolist(), 'gradient'] = 255\n",
    "            df['gradient'] = df['gradient'].apply(lambda x : 'rgb(255,' + str(255 - x) + ',' + str(255 - x)+')')\n",
    "            del df['Position']\n",
    "            df = df[['Coordinates','Condition(s)','Accession','Variation', 'gradient']]\n",
    "            cols = ('Coordinates','Condition(s)','Accession','Variation')\n",
    "            df['link'] = 'https://www.ncbi.nlm.nih.gov/clinvar/variation/' + df['Accession']\n",
    "            print(df)\n",
    "            children.append(html.Table(\n",
    "                [html.Tr([html.Th(col) for col in cols], style = {'textAlign' : 'center', 'border': '1px solid black', 'border-collapse' : 'collapse'})] +\n",
    "                [html.Tr([\n",
    "                    html.Td(df.iloc[i][col], style = {'backgroundColor' : df.iloc[i]['gradient'], 'textAlign' : 'center', 'border': '1px solid black', 'border-collapse' : 'collapse'}) if col != 'Accession' else html.Td(html.A(href=df.iloc[i]['link'], children=df.iloc[i][col], target='_blank'), style = {'backgroundColor' : df.iloc[i]['gradient'], 'textAlign' : 'center', 'border': '1px solid black', 'border-collapse' : 'collapse'}) for col in cols\n",
    "                ]) for i in range(len(df))],\n",
    "                style = {'textAlign' : 'center', 'border': '1px solid black', 'border-collapse' : 'collapse'}\n",
    "            )\n",
    "        )\n",
    "    except:\n",
    "        children.append(html.P('Could not find pathogenic variants for this gene in ClinVarðŸ¤· ðŸ¤·â€â™‚ï¸'))\n",
    "    return children"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "external_stylesheets = ['assets/BAMdelbee_main.css']\n",
    "app = dash.Dash(__name__, external_stylesheets=external_stylesheets)\n",
    "app.title = \"BAMdelbee | Ohad Birk's Lab\"\n",
    "app.config['suppress_callback_exceptions'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "app.layout = html.Div([\n",
    "    html.Div([\n",
    "        html.Img(id = 'header_image',\n",
    "            src = 'assets/logo_birklab.png',\n",
    "            style = {\n",
    "                'height' : '25%',\n",
    "                'width' : '25%',\n",
    "                'float' : 'left'\n",
    "            }\n",
    "        ),\n",
    "        html.Span(''),\n",
    "        html.Img(id = 'header_image2',\n",
    "            src = 'assets/logo_birklab.png',\n",
    "            style = {\n",
    "                'height' : '25%',\n",
    "                'width' : '25%',\n",
    "                'float' : 'right',\n",
    "                'margin' : 4\n",
    "            }\n",
    "        ),\n",
    "    ]),\n",
    "    html.Span(),\n",
    "    html.H2(\n",
    "        'Select a batch',\n",
    "        id = 'instructions',\n",
    "        style = {'width' : '100%', 'height' : '100%', 'font-family' : 'gisha', 'marginLeft' : 'auto', 'marginRight' : 'auto', 'textAlign' : 'center'}\n",
    "\n",
    "    ),\n",
    "    dcc.Dropdown(\n",
    "        id = 'BatchesDropdown',\n",
    "        options = [{'label': v[8:-4], 'value': v} for v in batches],\n",
    "        multi = False,\n",
    "        placeholder = 'Select batch',\n",
    "        style = default_style\n",
    "    ),\n",
    "    dcc.RadioItems(\n",
    "        id = 'referenceRadioButtons',\n",
    "        options=[\n",
    "            {'label': 'hg38', 'value': 'hg38'},\n",
    "            {'label': 'hg19', 'value': 'hg19'},\n",
    "        ],\n",
    "        value='hg38',\n",
    "        labelStyle={'display': 'inline-block'},\n",
    "        style = default_style\n",
    "    ),\n",
    "    html.Div([\n",
    "        html.Div([\n",
    "            html.Div(id = 'checkListContainer'),\n",
    "        ], id = 'checklist_div',\n",
    "        style = {'display' : 'None'}\n",
    "        )],\n",
    "        id = 'parameters_div',\n",
    "        style = default_style\n",
    "    ),\n",
    "    html.Div(\n",
    "        dcc.Loading(\n",
    "            id = 'loading_results',\n",
    "            fullscreen = True,\n",
    "            children = [html.Div([html.Div(id = 'loading_table_results')])],\n",
    "            type = 'circle',\n",
    "            style = {'marginTop' : '120px'}\n",
    "        ),\n",
    "        id = 'loading_results_div',\n",
    "        style={'width': '100%', 'height': '10%', 'borderWidth': '1px', 'borderStyle': 'grooved', 'borderRadius': '5px', 'textAlign': 'center', 'marginLeft':'auto', 'marginRight':'auto'}\n",
    "    ),\n",
    "    html.Div(\n",
    "        dcc.Loading(\n",
    "            id = 'loading_heatTable',\n",
    "            fullscreen = True,\n",
    "            children = [html.Div([html.Div(id = 'loading_heattable_results')])],\n",
    "            type = 'circle',\n",
    "            style = {'marginTop' : '120px'}\n",
    "        ),\n",
    "        id = 'loading_heatTable_div',\n",
    "        style={'width': '100%', 'height': '10%', 'borderWidth': '1px', 'borderStyle': 'grooved', 'borderRadius': '5px', 'textAlign': 'center', 'marginLeft':'auto', 'marginRight':'auto'}\n",
    "    ),\n",
    "    html.Div(id = 'table_div', style = hide)\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@app.callback(\n",
    "    [Output('BatchesDropdown', 'style'),\n",
    "     Output('referenceRadioButtons', 'style'),\n",
    "     Output('checklist_div', 'style'),\n",
    "     Output('instructions', 'children'),\n",
    "     Output('checkListContainer', 'children')],\n",
    "    [Input('BatchesDropdown', 'value')]\n",
    ")\n",
    "def chooseBatch(batch):\n",
    "    if batch != None:\n",
    "        zBatch = ZipFile(batch)\n",
    "        samples = zBatch.read('results/samples.txt').decode('utf-8', 'ignore').split('\\n')\n",
    "        samples = [{'label' : str(sample), 'value' : str(sample)} for sample in samples]\n",
    "        checkList = dcc.Checklist(id = 'samplesCheckList', options = samples, labelStyle={'display': 'block'}) \n",
    "        btn = html.Button('Analyze', id = 'Analysis_starting_button', disabled = True, n_clicks = 0, style = {'text-transform' : 'none', 'font-size' : '30px'})\n",
    "        return [hide, hide, default_style, batch[8:-4], [checkList, btn]]\n",
    "    else:\n",
    "        return [default_style, default_style, hide, \"Select batch:\", None]\n",
    "\n",
    "@app.callback(\n",
    "    [Output('Analysis_starting_button', 'disabled')],\n",
    "    [Input('samplesCheckList', 'value')]\n",
    ")\n",
    "def chooseSamples(samples):\n",
    "    if samples not in [None, []]:\n",
    "        return [False]\n",
    "    else:\n",
    "        return [True]\n",
    "\n",
    "@app.callback(\n",
    "    [Output('parameters_div', 'style'),\n",
    "     Output('table_div', 'style'),\n",
    "     Output('table_div', 'children'),\n",
    "     Output('loading_results_div', 'children')],\n",
    "    [Input('Analysis_starting_button', 'n_clicks'),\n",
    "     Input('BatchesDropdown', 'value'),\n",
    "     Input('samplesCheckList', 'value'),\n",
    "     Input('referenceRadioButtons', 'value')]\n",
    ")\n",
    "def update_figure(clicks, batch, affected, referenceGenome):\n",
    "    if clicks in (0, None):\n",
    "        return [default_style, hide, [], None]\n",
    "    else:\n",
    "        if referenceGenome == 'hg38':\n",
    "            geneMap = pd.read_csv('Databases/NCBI_Genes_hg38.map')\n",
    "        else:\n",
    "            geneMap = pd.read_csv('Databases/NCBI_Genes_hg19.map')\n",
    "        geneMap['Chr'] = geneMap['Chr'].astype(str)\n",
    "        \n",
    "        zBatch = ZipFile(batch)\n",
    "        deletions = []\n",
    "        sizes = []\n",
    "        genes = []\n",
    "        \n",
    "        for chromosome in [str(c) for c in range(1,23)] + ['X', 'Y']:\n",
    "            try:\n",
    "                df = pd.read_csv(zBatch.open('results/chr' + chromosome + '.coverage'), sep = '\\t')\n",
    "            except:\n",
    "                continue\n",
    "            samples = df.columns[1:]\n",
    "            controls = [sample for sample in samples if sample not in affected]\n",
    "            for sample in affected:\n",
    "                df = df[df[sample] == 0]\n",
    "            for control in controls:\n",
    "                df = df[df[control] > 5]\n",
    "            if df.shape[0] < 2:\n",
    "                continue\n",
    "            coordinates = df['Coordinate'].tolist()\n",
    "            while len(coordinates) != 0:\n",
    "                for coordinate in coordinates:\n",
    "                    start = coordinate\n",
    "                    end = coordinate\n",
    "                    while end + 50 in coordinates:\n",
    "                        end = end + 50\n",
    "                    coordinates = [coordinate for coordinate in coordinates if coordinate > end]\n",
    "                    if end != start:\n",
    "                        deletions.append(chromosome + ':' + str(start) + '-' + str(end))\n",
    "                        sizes.append(end-start)\n",
    "                        genes_in_deletion = geneMap[(geneMap['Chr'] == chromosome) & (geneMap['Start'] <= start) & (geneMap['End'] >= end)]['Gene'].tolist()\n",
    "                        genes.append(', '.join(genes_in_deletion))\n",
    "        df = pd.DataFrame()\n",
    "        df['Coordinates'] = deletions\n",
    "        df['Chromosome'] = df['Coordinates'].apply(lambda x : int(x.split(':')[0].upper().replace('X', '23').replace('Y', '24').replace('M', '25').replace('MT', '25')))\n",
    "        df['~Size (bp)'] = sizes\n",
    "        df['Genes'] = genes\n",
    "        df['Start'] = df['Coordinates'].apply(lambda x : int(x.split(':')[1].split('-')[0]))\n",
    "        df.sort_values(['Chromosome', 'Start'], inplace = True)\n",
    "        df.reset_index(drop = True, inplace = True)\n",
    "        df['End'] = df['Coordinates'].apply(lambda x : int(x.split(':')[1].split('-')[1]))\n",
    "        del df['Chromosome']\n",
    "        del df['Start']\n",
    "        del df['End']\n",
    "        df = df[df['~Size (bp)'] > 50]\n",
    "\n",
    "        ddt = dash_table.DataTable(\n",
    "            id = 'table',\n",
    "            data = df.to_dict('records'),\n",
    "            columns = [{'id': c, 'name': c} for c in df.columns],\n",
    "            page_action = 'none',\n",
    "            row_selectable='single',\n",
    "            fixed_rows={'headers' : True},\n",
    "            style_cell = {\n",
    "                'textAlign': 'left',\n",
    "                'overflow': 'hidden',\n",
    "                'textOverflow': 'ellipsis',\n",
    "                'maxWidth': 0\n",
    "            },\n",
    "            tooltip_delay=0,\n",
    "            tooltip_duration=None,\n",
    "\n",
    "            style_data_conditional=[\n",
    "                {\n",
    "                'if': {'row_index': 'odd'},\n",
    "                'backgroundColor': 'rgb(247, 247, 255)'\n",
    "                },\n",
    "                {\n",
    "                'if': {'row_index': 'even'},\n",
    "                'backgroundColor': 'rgb(252, 252, 255)'\n",
    "                },\n",
    "            ],\n",
    "            style_data={\n",
    "                'whiteSpace': 'normal',\n",
    "                'height': 'auto',\n",
    "                },\n",
    "            style_header = {\n",
    "                'backgroundColor': 'rgb(220, 220, 255)',\n",
    "                'fontWeight': 'bold',\n",
    "                'whiteSpace' : 'normal'\n",
    "                },\n",
    "            style_table={\n",
    "                'maxHeight': '250px',\n",
    "                'maxwidth' : '150%',\n",
    "                'border': 'thin lightgrey solid'\n",
    "            },\n",
    "            style_as_list_view = False\n",
    "        )\n",
    "        info = html.Div('Click on a deletion for more information', id = 'info')\n",
    "        return [hide, full_display, [ddt, info], None]\n",
    "\n",
    "@app.callback(\n",
    "    [Output('info', 'children'),\n",
    "     Output('info', 'style')],\n",
    "    [Input('table', 'derived_virtual_selected_rows'),\n",
    "     Input('table', 'derived_virtual_data'),\n",
    "     Input('referenceRadioButtons', 'value')]\n",
    ")\n",
    "def getVariantData(selected_row_index, data, referenceGenome):\n",
    "    if selected_row_index in [[], None]:\n",
    "        return [html.P('Select row for more information'), default_style]\n",
    "    else:\n",
    "        elements = []\n",
    "        coordinates = data[selected_row_index[0]]['Coordinates']\n",
    "        genes = data[selected_row_index[0]]['Genes']\n",
    "        \n",
    "        #IGV & UCSC % gnomAD\n",
    "        igv_link = 'http://localhost:60151/goto?locus=chr' + coordinates\n",
    "        igvA = html.A('IGV', href = igv_link, target = '_blank')\n",
    "        ucsc_link = 'https://genome.ucsc.edu/cgi-bin/hgTracks?db=' + referenceGenome + '&position=' + coordinates.replace(':','%3A')\n",
    "        ucscA = html.A('UCSC', href = ucsc_link, target = '_blank')\n",
    "        if referenceGenome == 'hg38':\n",
    "            gnomAD_Url = 'https://gnomad.broadinstitute.org/region/' + coordinates + '?dataset=gnomad_r3'\n",
    "        else:\n",
    "            gnomAD_Url = 'https://gnomad.broadinstitute.org/region/' + coordinates + '?dataset=gnomad_r2_1'\n",
    "        gnomADLink = html.A('gnomAD',target='_blank', href = gnomAD_Url)\n",
    "        elements.append(html.Div([igvA, html.Span('    '), ucscA, html.Span('    '), gnomADLink]))\n",
    "        \n",
    "        if len(genes) > 0:\n",
    "            \n",
    "            #RefSeq\n",
    "            for gene in genes.split(', '):\n",
    "                try:\n",
    "                    refseq = refSeqSummaries[refSeqSummaries['Symbol'] == gene]\n",
    "                    fullName = refseq['Full Name'].tolist()[0]\n",
    "                    elements.append(html.Div(fullName, style = {'fontStyle' : 'oblique'}))\n",
    "                    summary = refseq['Summary'].tolist()[0]\n",
    "                    elements.append(html.Div(summary))\n",
    "                    elements.append(break_line)\n",
    "                except: False\n",
    "        \n",
    "            #HPO\n",
    "            hpo_terms = hpo[hpo['Associated Genes'].str.contains('|'.join(genes.split(', ')))]['HPO'].tolist()\n",
    "            phenotypes = ', '.join(hpo_terms)\n",
    "            if len(phenotypes.strip()) > 0:\n",
    "                elements.append(html.Div('HPO associated phenotypes:', style = {'fontStyle' : 'oblique'}))\n",
    "                elements.append(html.Div(phenotypes))\n",
    "                elements.append(break_line)\n",
    "        \n",
    "            #GeneCards\n",
    "            geneCards = [html.Span('GeneCards: ')]\n",
    "            for gene in genes.split(', '):\n",
    "                geneCards_link = 'https://www.genecards.org/cgi-bin/carddisp.pl?gene=' + gene\n",
    "                geneCardsA = html.A(gene, href = geneCards_link, target = '_blank')\n",
    "                geneCards.append(html.Span('    '))\n",
    "                geneCards.append(geneCardsA)\n",
    "            elements.append(html.Div(geneCards))\n",
    "        \n",
    "            #OMIM\n",
    "            omims = [html.Span('OMIM: ')]\n",
    "            for gene in genes.split(', '):\n",
    "                try:\n",
    "                    omim = str(omim_map[omim_map['Approved Gene Symbol (HGNC)'] == gene]['MIM Number'].tolist()[0]).replace('.0','')\n",
    "                    omim_link = 'https://omim.org/entry/' + omim\n",
    "                    omimA = html.A(gene, href = omim_link, target = '_blank')\n",
    "                    omims.append(html.Span('    '))\n",
    "                    omims.append(omimA)\n",
    "                except: False\n",
    "            if len(omims) > 1:\n",
    "                elements.append(html.Div(omims))\n",
    "        \n",
    "            #GTEx\n",
    "            GTExs = [html.Span('GTEx: ')]\n",
    "            for gene in genes.split(', '):\n",
    "                GTEx_link = 'https://gtexportal.org/home/gene/' + gene\n",
    "                GTExA = html.A(gene, href = GTEx_link, target = '_blank')\n",
    "                GTExs.append(html.Span('    '))\n",
    "                GTExs.append(GTExA)\n",
    "            elements.append(html.Div(GTExs))\n",
    "\n",
    "            #MGI\n",
    "            mgis = [html.Span('MGI: ')]\n",
    "            for gene in genes.split(', '):\n",
    "                try:\n",
    "                    mgi_link = 'http://www.informatics.jax.org/marker/phenotypes/MGI:' + MGI_IDs[gene]\n",
    "                    mgiA = html.A(gene, href = mgi_link, target = '_blank')\n",
    "                    mgis.append(html.Span('    '))\n",
    "                    mgis.append(mgiA)\n",
    "                except: False\n",
    "            if len(mgis) > 1:\n",
    "                elements.append(html.Div(mgis))\n",
    "        \n",
    "        #ClinVar HeatTable\n",
    "        if len(genes) > 0:\n",
    "            elements.append(html.Div(break_line))\n",
    "            clinVarBtn = btn = html.Button('ClinVar HeatTable', id = 'HeatTableBtn', n_clicks = 0, style = {'text-transform' : 'none'})\n",
    "            heatTableDiv = html.Div(id = 'heatTableDiv')\n",
    "            elements.append(clinVarBtn)\n",
    "            elements.append(heatTableDiv)\n",
    "        \n",
    "        return [elements, info_style]\n",
    "\n",
    "@app.callback(\n",
    "    [Output('heatTableDiv', 'children'),\n",
    "     Output('loading_heatTable_div', 'children'),\n",
    "     Output('HeatTableBtn', 'style')],\n",
    "    [Input('HeatTableBtn', 'n_clicks'),\n",
    "     Input('table', 'derived_virtual_selected_rows'),\n",
    "     Input('table', 'derived_virtual_data')]\n",
    ")\n",
    "def update_figure(clicks, selected_row_index, data):\n",
    "    if clicks in [0, None]:\n",
    "        return [[None], None]\n",
    "    else:\n",
    "        coordinates = data[selected_row_index[0]]['Coordinates']\n",
    "        gene = data[selected_row_index[0]]['Genes']\n",
    "        if ', ' in gene:\n",
    "            gene = gene.split(', ')[0]\n",
    "        return [generateHeatTable(gene, coordinates.split(':')[0], coordinates.split(':')[1].split('-')[0]), None, {'display' : 'None'}]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    import webbrowser\n",
    "    webbrowser.open('http://127.0.0.1:6161/')\n",
    "    app.run_server(port=6161)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
